torch/distributed/checkpoint/state_dict_saver.py

```diff
@@ -212,9 +215,58 @@ def async_save(
             torch.device("cpu") in pg._device_types  # type: ignore[attr-defined]
         ), "A CPU backend must be enabled for async save; try initializing process group with 'cpu:gloo,cuda:nccl'"
 
-    cpu_state_dict = _offload_state_dict_to_cpu(_stateful_to_state_dict(state_dict))
-
-    executor = ThreadPoolExecutor(max_workers=1)
+    # This is the simplest implementation of a "stage" specific method which occurs pre-execution
+    # This is done in a way that is BC, and does not force the creation of `stage_state_dict` for
+    # users that don't really want to think about this
+
+    # Issue #1
+    # `staging` still has no explicit path to the same local/global planning
+    # via the usual planner interface. If we want to do things like dedup replicas to avoid
+    # extraneous D2H pre- submitting the save job, we still don't have an (explicit) way of handling that
+    # in this proposal. (Users could potentially call `dcp.save inside their custom 'stage' method`)
+
+    # Issue #2
+    # Users are forced to use a thread as the async vehicle, or provide some ThreadPoolExecutor-like object
+    # Is there a cleaner way to implement this without relying on ThreadPoolExecutor? Do we want to support this?
+    # This may not be true for users which hack storage_writer and planner methods to be async under the hood, but
+    # IMO the UX is shaky.
+    # This might be important in xlformers (eventually) or in cases
+    # where users are worried about GIL issues.
+
+    # A totally reasonable alternative to this entire proposal is to expect users who want extra
+    # customization to implement their own async_save methods, and rely on internal calls to dcp.save
+    # in fact even with these extensions, this might still be what we tell users to do if this isn't
+    # "good enough"
+    # here's once example which accomplishes this and addresses both issues
+
+
+    # def custom_async_save(...):
+    #     this step accomplishes staging and includes the usual 'planning' calls (issue 1)
+    #     buffered_writer = CpuBufferedWriter() # this is stateful, contains a copy of state_dict
+    #     dcp.save(state_dict, storage_writer=buffered_writer)
+
+    #     final_storage_writer = FileSystemWriter()
+    #     create whatever thread you want (issue #2)
+    #     mp.spawn(
+    #         dcp.save,    # or some custom sub-process method wich calls dcp.save under the hood
+    #         save,
+    #         buffered_writer.state_dict,   # lot's of way's to do this, not really the most important part
```

Comment:
Can you elaborate on the state_dict type here. Usually state_dict means, materialized data on a memory device. However, with some staging devices, we may not have luxury of having the staged data copy in-memory. Therefore, IMHO this is a very important API to expand on.
Can I expect the buffered_writer.state_dict to be of type Dict[str, Any] ?

Reply:
Sorry, for the noise. StateDictType is already support Any. Resolving.